js/redux/modules/teams.js

```diff
@@ -0,0 +1,202 @@
+  type: GET_TEAM_ERROR,
+  payload: error
+});
+const getTeamLoading = () => ({
+  type: GET_TEAM_LOADING
+});
+
+export const getTeamReset = () => ({
+  type: GET_TEAM_RESET
+});
+
+// Async action creators
+
+// Async action to fetch best match to user
+export const fetchBestMatch = (
+  workstyle,
+  score,
+  competitionId,
+  teamSize
+) => async dispatch => {
+  dispatch(getTeamReset());
+  dispatch(getTeamLoading());
+
+  const type = teamType(...score);
+  const teams = {};
+
+  const teamsQuery = firebaseDB
+    .collection("teams")
+    .where("competitionId", "==", competitionId)
+    .where("workstyle", "==", workstyle)
+    .where("type", "==", type);
+
+  await teamsQuery
+    .get()
+    .then(snapshot => {
+      if (!snapshot.empty) {
+        snapshot.forEach(team => {
+          teams[team.id] = team.data();
+        });
+      }
+    })
+    .catch(error => dispatch(getTeamError(error)));
+
+  if (Object.keys(teams).length) {
+    const { match, otherMatches } = findBestMatch(teams);
+    dispatch(getTeamBestMatch(match));
+    dispatch(getTeamOtherMatches(otherMatches));
+  } else {
+    // if there's no match with the user type, fetch similar type teams
+    const rematch = await dispatch(
+      fetchOtherMatches(workstyle, score, competitionId, true)
+    );
+    const result = findBestMatch(rematch);
+
+    // if there's no match with the similar types either
+    // create new team
+    if (result.match === undefined) {
+      const uid = await AsyncStorage.getItem("user");
+      firebaseDB
+        .collection("teams")
+        .add({
+          workstyle,
+          type,
+          competitionId,
+          teamSize,
+          users: {
+            uid: { fun: score[0], grow: score[1], win: score[2], id: uid }
+          }
+        })
+        .then(async docRef => {
+          // Refetch newly added team, which is a perfect match
+          const newTeam = await docRef.get();
+          dispatch(getTeamBestMatch(newTeam.data()));
+        })
+        .catch(err => dispatch(getTeamError(err)));
+    } else {
+      dispatch(getTeamBestMatch(result.match));
+      dispatch(getTeamOtherMatches(result.otherMatches));
+    }
+  }
+};
+
+// Async action to fetch other matches when user clicks "browse other groups"
+// @param: user workstyle, score, competition id and flag true for refetch if no best match exists
+export const fetchOtherMatches = (
+  workstyle,
+  score,
+  competitionId,
+  flag
+) => async dispatch => {
+  dispatch(getTeamLoading());
+
+  const [typeOne, typeTwo] = similarTeamType(...score);
+  const matches = [];
+
+  const teamsQueryOne = firebaseDB
+    .collection("teams")
+    .where("competitionId", "==", competitionId)
+    .where("workstyle", "==", workstyle)
+    .where("type", "==", typeOne);
+
+  const teamsQueryTwo = firebaseDB
+    .collection("teams")
+    .where("competitionId", "==", competitionId)
+    .where("workstyle", "==", workstyle)
+    .where("type", "==", typeTwo);
+
```

Comment:
I think all your DB queries can go to a helper file.
You can write a generic function and reuse it as well.